src/python/pants/engine/rules.py

```diff
@@ -252,30 +264,50 @@ class RuleGraph(datatype('RuleGraph',
   Because in
 
      `root_subject_types` the root subject types this graph was generated with.
-     `root_rules` The rule entries that can produce the root products this graph was generated
-                        with.
+     `root_rules` A map from root rules, ie rules representing the expected selector / subject types
+                  for requests, to the rules that can fulfill them.
      `rule_dependencies` A map from rule entries to the rule entries they depend on.
                          The collections of dependencies are contained by RuleEdges objects.
                          Keys must be subclasses of CanHaveDependencies
                          values must be subclasses of CanBeDependency
      `unfulfillable_rules` A map of rule entries to collections of Diagnostics
-                                 containing the reasons why they were eliminated from the graph.
+                           containing the reasons why they were eliminated from the graph.
 
   """
-  # TODO constructing nodes from the resulting graph.
-  # Possible approach:
-  # - walk out from root nodes, constructing each node.
-  # - when hit a node that can't be constructed yet, ie the subject type changes,
-  #   skip and collect for later.
-  # - inject the constructed nodes into the product graph.
+
+  def dependency_edges_for_rule(self, rule, subject_type):
+    if type(rule) is RootRuleGraphEntry:
+      return self.root_rule_edges(RootRuleGraphEntry(rule.subject_type, rule.selector))
+    else:
+      return self.rule_dependencies.get(RuleGraphEntry(subject_type, rule))
+
+  def is_unfulfillable(self, rule, subject_type):
+    return RuleGraphEntry(subject_type, rule) in self.unfulfillable_rules
+
+  def root_rule_matching(self, subject_type, selector):
+    root_rule = RootRuleGraphEntry(subject_type, selector)
+    if root_rule in self.root_rules:
+      return root_rule
+
+  def new_graph_with_root_for(self, subject_type, selector):
```

Comment:
Yeah. I noted that in the description, but I don't think it was sufficiently called out. I could just rm them since I'll be reimplementing in rust anyway.